zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServer.java

```diff
@@ -1151,6 +1213,9 @@ protected void setLocalSessionFlag(Request si) {
     }
 
     public void submitRequest(Request si) {
+        if (state == State.MAINTENANCE) {
+            throw new IllegalStateException("Zookeeper server is in maintenance state");
+        }
```

Comment:
If you throw an exception here, the request will fail on the client side. I haven't tried it myself, have you valdiated that the client is able to seemlessly retry the command without bothering the user?
If not, I suggest another solution: create a semaphore which is closed while the maintenace is happening and block this thread until it's finished. The request will suffer some additionaly latency, but otherwise cannot be noticed. Thoughts?

Reply:
blocking here is probably not a good idea, we risk to block the pipeline, with unpredictable consequences probably

Reply:
Moreover, request throttler does the throttling mechanism by blocking this thread.

Reply:
you are right @anmolnar 
thanks for your clarification.

we could have a similar loop here

```
while (state == State.MAINTENANCE) {
    wait(1000);
}
```

Reply:
Thanks for the inputs. I will  look into this.

Reply:
Yeah, be aware that you have to do `notifyAll()` after changing the state back to RUNNING to notify the waiting threads. I'll add another comment at the right place.

Additionally, I don't get this `check-wait(timeout)-check-wait(timeout)-...` logic. It doesn't make sense to me, there's no additional fail-safe logic inside the loop, why don't just `wait()`? Instead, it releases the lock in every second probably without any benefit. I don't say we have to change that, I just don't understand.

Reply:
> If you throw an exception here, the request will fail on the client side. I haven't tried it myself, have you valdiated that the client is able to seemlessly retry the command without bothering the user?

I checked the scenario of sending client request while zk is in maintenance state. Here is what I found.

1. Connection/session creation will be retried but read/write operation will not.
2. The read/write request will fail even no exception is thrown here, because the request fails before sending to ZookeeperServer. Here is what happens:

a)  `ServerCnxn` detects that ZK server is not running and close the socket/channel. 
b) `ClientSocketCnxn.doTransport()` then detects that socket/channel is closed and throws IOException
c)  'SendThread' catches the Exception, adds `ConnectionLoss` error reply header, clears up the outgoingQueue ann pendingQueue, and attempts to re-connect to server
d)  Zookeeper check the header and sends ConnectionLoss to user

Reply:
Makes sense to me.

Reply:
Got you. There're 2 things that crossed my mind to resolve this:
1. Add MAINTENANCE as a running state to `ZooKeeperServer.isRunning()` method.
2. Do not alter the state of ZooKeeperServer, but use a separate volatile variable in this class to block requests temporarly.

Reply:
Thanks for the feedback and inputs, @eolivelli  and @anmolnar

Reply:
> Add MAINTENANCE as a running state to ZooKeeperServer.isRunning() method.

Adding MAINTENANCE to ZooKeeperServer.isRunning() will change the semantics of "running" state and has bigger impact, we would want to avoid it if it's possible.

> Do not alter the state of ZooKeeperServer, but use a separate volatile variable in this class to block requests temporarly.

Good idea. Yes, we don't have to alter the state of ZooKeeper if we don't think we need a generic MAINTENANCE state. 

Instead of using volatile variable,  I will add a `CountDownLatch` to efficiently coordinate/sync between threads and utilize system resources. I will update the  PR with the changes shortly.

Reply:
Makes sense to me.
Maybe  before starting to way you can log at INFO level a line like "MAINTEINANCE in progress, blocking the request processing"

Reply:
+1 for the logging suggestion of @eolivelli .
I also like the Latch approach @li4wang . I think you have to use `AtomicReference<CountDownLatch>` for that, because you create a new instance every time ZK enters into Maintenance mode. In `submitRequest()` you'll try grabbing the reference by `get()` and if it's not null, await for it.

Reply:
How about using `volatile` instead of `AtomicReference` to keep it simple and lightweight? In this case, we only need to take care of visibility issue. Only one thread writes the value, so no need for the additional power (i.e. atomicity/CAS) provided by AtomicReference.

With `volatile`,  we can ensure  once the restore thread creates a new instance, the instance value will be immediately visible to threads that check if it's null.

Reply:
updated